### ISSUE TITLE
visual_studio generator's conanbuildinfo.props missing ResourceCompile

### ISSUE THREAD

**Issue by justinpa-asg** [conan-io/conan#1421]
My Visual Studio project builds resource files that cannot include header files. I had to copy "<ClCompile><AdditionalIncludeDirectories>...</ClCompile>" to "<ResourceCompile>...</ResourceCompile>". Can the latter be added by Conan's visual_studio generator?

In the near-term, how do I easily modify or create a new generator? Can I derive from the existing generator? Or should I copy and add?

**Comment by memsharded** [comment]
Hi Justin,

If the modification is general enough for most Visual Studio users, of course we can change it, or you can contribute with a pull requests with the desired functionality.

For testing in the near term, you have 2 approaches:
- Fork the conan codebase and run from sources with your changes.
- Create your own generator ``visual_studio2`` in the form of a package and "require" it to be able to use it. You have details in: http://docs.conan.io/en/latest/howtos/dyn_generators.html. You could copy the initial code from the conan codebase.

I would like to understand the issue, not an expert in the resource files of VS, what did you have to copy from where to where? If you could elaborate a bit further, please. 
Tell me if you need help with any of the above approaches. Thanks!

**Comment by justinpa-asg** [comment]
file:~/.conan/data/MyPackage/1.2/user/channel/build/28..d6/conanbuildinfo.props
```
  <ItemDefinitionGroup>
    <ClCompile>
      <AdditionalIncludeDirectories>~/.conan/data/MyOtherPackage/1.4/user/channel/package/fa..d7/include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
      <PreprocessorDefinitions>%(PreprocessorDefinitions)</PreprocessorDefinitions>
      <AdditionalOptions> %(AdditionalOptions)</AdditionalOptions>
    </ClCompile>
+    <ResourceCompile>
+      <AdditionalIncludeDirectories>~/.conan/data/MyOtherPackage/1.4/user/channel/package/fa..d7/include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <PreprocessorDefinitions>%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <AdditionalOptions> %(AdditionalOptions)</AdditionalOptions>
+    </ResourceCompile>
```

**Comment by memsharded** [comment]
So that is what should be modified, good.
But why? Sorry my ignorance about the visual studio properties files. What kind of files inside the "include" folders will be processed with this? What does VS do with those files? Could that affect existing users of the generator, any risk to break other users projects using this generator? Thanks!

**Comment by justinpa-asg** [comment]
I exported a new recipe with class visual_studio2(Generator) and added per documentation.

```
    requires = "ConanGenerators/1.0@user/channel"
    generators = "visual_studio2", "gcc", "txt"
```

I'm getting:

> MyProject/1.2@user/channel: WARN: Invalid generator 'visual_studio2'. Available types: cmake_multi, visual_studio, gcc, qbs, cmake, env, xcode, virtualenv, qmake, virtualbuildenv, scons, ycm, virtualrunenv, txt

Without the ResourceCompile section in properties file, I get:

> .conan\data\MyProject\1.2\user\channel\build\02..98\MyPackage\config\VisualStudio2015\MyPackage.sln" (default target) (1) ->
> .conan\data\MyPackage\1.2\user\channel\build\02..098\MyProject\config\VisualStudio2015\MyProject.vcxproj" (default target) (26) ->
>   ..\..\src\main\rc\MyProject.rc(10): fatal error RC1015: cannot open include file 'myproject.hh'. [.conan\data\MyProject\1.2\user\channel\build\02..98\MyProject\config\VisualStudio2015\MyProject.vcxproj]

I'm not a Visual Studio expert either, but I assume ResourceCompile section will be ignored by projects without resource files.

**Comment by memsharded** [comment]
Don't know what could be happening, would like to have a look to the full generator. Could you please put it in a github repo, so I can check it? Thanks!

**Comment by justinpa-asg** [comment]
Attached.
[conanfile.txt](https://github.com/conan-io/conan/files/1096495/conanfile.txt)

**Comment by memsharded** [comment]
Ok, there is a small bug here! The name of dynamic generator cannot have a "_" in it. Will try to fix it for next release, in the meantime, would you mind to use a name for it without underscore?

Thanks very much for helping to catch it!

**Comment by memsharded** [comment]
The minor bug with the _ has been fixed (in develop branch) it will be released in next 0.25. So I have removed the "bug" label, but hasn't annotated as "fixed", because this is mainly a feature request about the ``ResourceCompile`` section.

I am trying to read and learn about this, (the documentation so far is quite bad). I understand that there are some resources that have to be compiled by ``rc.exe`` to be included in the project. My doubt here is why those resources would be in the ``include`` folders. Conan defined a resources directory (``cpp_info.resdirs``) that was defined to store generic resource files, like images, sounds, data files, configuration, etc. It seems that adding it to the compilation of the current package/project is not necessarily a good default.

Are you putting resource files inside include directories? Which kind of files? What is the result of not adding them to a ``ResourceCompile`` section?

**Comment by justinpa-asg** [comment]
My colleague tells me that these are standard include files generated by Visual Studio with pre-processor definitions. The *.rc then use those definitions. If I were to set self.cpp_info.resdirs instead of creating a custom generator, how would I set it to all the include directories of what the package requires?

```
    def package_info(self):
        if self.settings.os == "Windows":
            self.cpp_info.resdirs = ['requires1\include;requires2\include']
```

**Comment by justinpa-asg** [comment]
I tried:
`self.cpp_info.resdirs = self.deps_cpp_info['MyOtherPackage'].includedirs`

And still get:

> src\main\rc\MyProject.rc(10): fatal error RC1015: cannot open include file 'myproject.hh'.

I don't see any resource section in conanbuildinfo.props. Shouldn't resdirs propogate to this file?

**Comment by memsharded** [comment]
The idea was not to set the ``resdirs`` to point to the ``includedirs``, but trying to understand the issue and the flow.

> I don't see any resource section in conanbuildinfo.props. Shouldn't resdirs propogate to this file?

That is the question. Conan does not propagate it to a ``<ResourceCompile>`` section, because it had some problems. As the resources put in ``resdirs`` are any type of resources, not just Visual Studio resources, adding it to it IIRC was causing errors.

So lets go one step back, because I still don't get it, excuse my ignorance :)

- When you create a package, you put artifacts in it. You typically put the headers .h, .hpp, .hxx in the "include" folder, and the libraries .lib .a in the "lib" folder.
- These folders are added by default, or explicitely to the ``self.cpp_info`` object of the package.
- Consumers of the package will get an ``includedirs`` and ``libdirs`` paths, that will automatically be translated to their build system via generators, to variables indicating the include-path and the lib-path of the package, which is necessary to build and link consumers

So I would like to understand first what we are doing. From what I am reading in the Visual docs, it seems that such resources are designed to be consumed by the project (package) they are in, not the users/consumer of such project. Just using other package RC files might have issues, for example, you can get collisions of symbol IDs from different packages of your dependencies RC files (https://msdn.microsoft.com/en-us/library/c7h5sz68.aspx).

Is it the case that when building static libraries, the RC compiler generates a ``.res`` file that is not linked or included in the static library, and then the final executable or dll has to directly link with that ``.res`` artifact? It could seem that we could package the ``res`` artifact, but I guess that it might still suffer from the symbol collisions issue.

I have been trying to add some resources to the projects described here: http://blog.conan.io/2017/05/11/C-C++-Binary-Package-Management-for-Visual-Studio.html, but without much success, as I still don't understand the whole RC Visual studio feature.

It would be nice to setup a proof of concept, like adding some resource to: https://github.com/memsharded/hello_vs, and specifying why and how do we want to consume that resource from the consuming project: https://github.com/memsharded/chat_vs

**Comment by memsharded** [comment]
Maybe the feedback of other users using the ``visual_studio`` generator would be useful. 
Please, @davidkjaer, @youcefl, @Xaltotun, feedback very welcome regarding this ``ResourceCompile`` issue. Thanks very much!

**Comment by memsharded** [comment]
I am still missing some understanding of this feature, as 0.25 is about to be released, moving it forward to 0.26.

**Comment by justinpa-asg** [comment]
@memsharded Sorry, I'm not a Visual Studio developer. I'm just trying to build existing projects. While I prefer to use the default generator, the custom generator allows me to build by adding ResourceCompile Include directories. All I can contribute is that my colleague said resources were generated in Visual Studio. If resource files include headers, seems adding include directories is necessary.

**Comment by memsharded** [comment]
Yes, I am sorry too, I am really willing to help here, but I miss the necessary expertise, and I am afraid that adding that might break other users projects. Otherwise, I would add them. I will try to invest more time investigating this for next release. Thanks for your feedback!

**Comment by jnnnnn** [comment]
I would like Resource includes. It is possible to define a resource in a library or dll and then have your project.rc `#include` the resource ID, allowing you to use that resource in the dependent project.

For example, a library might have in `library.h`:

    #define ICON_ID_UNDO 9100
    #define ICON_ID_REDO 9101

then in `library.rc`:

    ICON_ID_UNDO ICON "res\\undo.ico"
    ICON_ID_REDO ICON "res\\redo.ico"

then in `executable.rc`:

    #include library.h

This allows dialogs in the executable to use the undo and redo icons from the library. The `#include` won't work if the ResourceCompile paths don't add the library's `include` directory. 

In our case, we have a library of UI controls (not just icons) that is used by hundreds of dependent projects.

The best documentation about this that I could find online was at [docs.microsoft.com](https://docs.microsoft.com/en-au/cpp/windows/including-shared-read-only-or-calculated-symbols).

**Comment by SSE4** [comment]
I think preprocessor definitions are needed as well

**Comment by memsharded** [comment]
I have added definitions and flags to ``<ResourceCompile>`` section: https://github.com/conan-io/conan/pull/2477
It seems it should break anything, but this is something difficult to know until released...

**Comment by memsharded** [comment]
This https://github.com/conan-io/conan/pull/2477 PR has been merged, will be released in 1.1, so I think this issue can be closed. Please re-open or comment if necessary.
